### PR TITLE
Add gradle plugin to send ProGuard files to Flurry

### PR DESCRIPTION
**Pull request by clayreimann**
Requires the shared code from #3

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -44,3 +46,27 @@ the _Run script only when installing` checkbox in the configuration.
 ```
 ./upload-symbols.py -c flurry.config -p ~/Downloads/dSYMs.zip
 ```
+
+## How to send ProGuard mapping files at build time
```
Comment (aklish) [inline README.md]: We should mention that our plugin depends on the android plugin

File: gradle/src/main/groovy/com/flurry/android/symbols/SymbolUploadPlugin.groovy [hunk 2]
```diff
@@ -0,0 +1,60 @@
+package com.flurry.android.symbols
```
Comment (clayreimann) [inline gradle/src/main/groovy/com/flurry/android/symbols/SymbolUploadPlugin.groovy]: added in c63f49c

File: gradle/src/main/groovy/com/flurry/android/symbols/SymbolUploadPlugin.groovy [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+class SymbolUploadPlugin implements Plugin<Project> {
+    public static final String CONFIGURATION_KEY = "flurryCrash"
+    public static final String API_KEY = "api-key"
+    public static final String TOKEN = "token"
+    public static final String FLURRY_UUID_KEY = "com.flurry.crash.map_id"
+
+    @Override
+    void apply(Project project) {
+        UploadProGuardMapping.logger = project.logger
+        UploadProGuardMapping.failHard = false
```
Comment (duseja2) [inline gradle/src/main/groovy/com/flurry/android/symbols/SymbolUploadPlugin.groovy]: Shouldn't the default here be true ?

File: gradle/src/main/groovy/com/flurry/android/symbols/SymbolUploadPlugin.groovy [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        getOrCreateConfig(project)
+
+        project.afterEvaluate {
+            SymbolUploadConfiguration config = getOrCreateConfig(project)
+            Map<String, String> configValues = evaluateConfig(config)
+
+            project.android.applicationVariants.all { BaseVariant variant ->
+                if (variant.mappingFile) {
+                    String uuid = UUID.randomUUID().toString()
```
Comment (duseja2) [inline gradle/src/main/groovy/com/flurry/android/symbols/SymbolUploadPlugin.groovy]: Does this mean we'll always use a random UUID ?

Comment (clayreimann) [inline gradle/src/main/groovy/com/flurry/android/symbols/SymbolUploadPlugin.groovy]: Yep, that's our current strategy.

File: gradle/src/main/groovy/com/flurry/android/symbols/SymbolUploadPlugin.groovy (hunk 3 above)
Comment (clayreimann) [inline gradle/src/main/groovy/com/flurry/android/symbols/SymbolUploadPlugin.groovy]: No, we want errors to bubble up. `failHard` exits the JVM on errors and is not a reasonable strategy for using the code as a library

Comment (clayreimann) [inline gradle/src/main/groovy/com/flurry/android/symbols/SymbolUploadPlugin.groovy]: On second thought, this should be the default behavior and the CLI should set `failHard` to `true`

File: gradle/src/main/groovy/com/flurry/android/symbols/SymbolUploadPlugin.groovy [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+        project.afterEvaluate {
+            SymbolUploadConfiguration config = getOrCreateConfig(project)
+            Map<String, String> configValues = evaluateConfig(config)
+            String apiKey = configValues[API_KEY]
+            String token = configValues[TOKEN]
+            int timeout = configValues[TIMEOUT].toInteger()
+
+            if (!apiKey) {
+                throw new IllegalStateException("You must provide the project's an API key")
```
Comment (aklish) [inline gradle/src/main/groovy/com/flurry/android/symbols/SymbolUploadPlugin.groovy]: grammar

File: README.md [hunk 6]
```diff
@@ -1,8 +1,10 @@
 # Flurry Upload Clients
 
 Flurry's Crash service can symbolicate the crashes reported by Flurry's SDK.
-This script uploads the symbols required to properly symbolicate crashes from
-iOS applications.
+This repository contains code to uploads the symbols required to properly symbolicate
+crashes from iOS apps and Android apps with ProGuard obfuscation. Both of these clients require
```
Comment (aklish) [inline README.md]: and deobfuscate Android apps with ProGaurd obfuscation.

File: proguard/src/main/java/com/flurry/proguard/UploadProguardMapping.java [hunk 7]
```diff
@@ -65,226 +83,350 @@ public static void main(String[] args) {
...
-        uploadFile(apiKey, uuid, pathToFile, token, timeout);
     }
 
+    /**
+     * Parses an INI-style config file
+     *
+     * @param filePath the path to the config file
+     * @return a map of the keys and values from the config
+     */
+    public static Map<String, String> parseConfigFile(String filePath) {
```
Comment (aklish) [inline proguard/src/main/java/com/flurry/proguard/UploadProguardMapping.java]: Can we use a Properties file here instead?

File: proguard/src/main/java/com/flurry/proguard/UploadProguardMapping.java [hunk 8]
```diff
@@ -65,226 +83,350 @@ public static void main(String[] args) {
...
+    private static void failWithError(String format, Object... args) {
+        LOGGER.error(format, args);
+        if (EXIT_PROCESS_ON_ERROR) {
+            System.exit(1);
+        } else {
+            String message = format;
+            Throwable cause = null;
+            if (args.length > 0 && args[args.length - 1] instanceof Throwable) {
+                cause = (Throwable) args[args.length - 1];
+                Object[] newArgs = new Object[args.length - 1];
```
Comment (aklish) [inline proguard/src/main/java/com/flurry/proguard/UploadProguardMapping.java]: this is unused.

File: proguard/src/main/java/com/flurry/proguard/UploadProguardMapping.java [hunk 9]
```diff
@@ -65,226 +83,350 @@ public static void main(String[] args) {
...
+                for (int i = 0; i < args.length - 1; i++) {
```
Comment (aklish) [inline proguard/src/main/java/com/flurry/proguard/UploadProguardMapping.java]: We can use Arrays.copyOf I believe here.

File: README.md (hunk 1 above)
Comment (clayreimann) [inline README.md]: Done in 0fdc157

**Comment by duseja2** [comment]
:+1: Feel free to merge, after your tests succeed

**Pull request merged**